test/gpu/native/studies/transpose/transpose.chpl

```diff
@@ -0,0 +1,98 @@
+use CTypes;
+use GPUDiagnostics;
+use GPU;
+use Time;
+
+config const numTrials = 10;
+config const useNaive = false;
+config const sizeX = 2048*8,
+             sizeY = 2048*8;
+config param blockSize = 16;
+
+pragma "codegen for GPU"
+pragma "always resolve function"
+export proc transposeMatrix(odata: c_ptr(real(32)), idata: c_ptr(real(32)), width: int, height: int) {
+  // Allocate extra columns for the shared 2D array to avoid bank conflicts.
+  var smVoidPtr = __primitive("gpu allocShared", numBytes(real(32))*(blockSize+1)*blockSize);
+  var smArrPtr = smVoidPtr : c_ptr(real(32));
+
+  const blockIdxX = __primitive("gpu blockIdx x"),
+        threadIdxX = __primitive("gpu threadIdx x"),
+        blockIdxY = __primitive("gpu blockIdx y"),
+        threadIdxY = __primitive("gpu threadIdx y");
+  var idxX = blockIdxX * blockSize + threadIdxX,
+      idxY = blockIdxY * blockSize + threadIdxY;
+
+  // Store the input data in transposed order into temporary array
+  // i.e., the below is effectively smArrPtr[y][x] instead of
+  // smArrPtr[x][y] for copy
+  if (idxX < width && idxY < height) {
+    smArrPtr[(blockSize+1) * threadIdxX + threadIdxY] = idata[idxY * width + idxX];
+  }
+
+  // synchronize the threads
+  __primitive("gpu syncThreads");
+
+  // Swap coordinates and write back out
+  idxX = blockIdxY * blockSize + threadIdxX;
+  idxY = blockIdxX * blockSize + threadIdxY;
+  if (idxX < height && idxY < width) {
+    odata[idxY * height + idxX] = smArrPtr[(blockSize+1) * threadIdxY + threadIdxX];
+  }
+}
+
+proc transposeTiled(original, output) {
```

Comment:
You use "tiled" here and "coalesced" elsewhere. Which are two optimizations you apply here as far as I can tell.

Could you use "low level" instead of these two? If someone bumps into this code, or the performance plot, I don't want to give them the impression this is how they're supposed to write optimized matrix transpose in Chapel. It can be even things like "with primitives" or something.